src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java

```diff
@@ -0,0 +1,130 @@
+package io.spokestack.spokestack.dialogue;
+
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The current state of the conversation at the time of a dialogue event.
+ */
+public final class ConversationState {
+
+    private final String frameName;
+    private final String nodeName;
+    private final String appAction;
+    private final Prompt systemPrompt;
+    private final String error;
+    private final Map<String, Object> slots;
+
+    /**
+     * Create a new conversation state.
+     *
+     * @param builder  The current conversation state builder.
+     */
+    private ConversationState(Builder builder) {
+        if (builder.conversationNode != null) {
+            String[] parts = builder.conversationNode.split("\\.", 2);
```

Comment:
This is awkward but I understand why it has to be like this

Reply:
Yeah, node naming is one of the things I'm on the fence about. It's not super-pretty.